File: Practice_Algorithm/77. Combinations.java
class Solution {
    public void addList(List<List<Integer>>res, int start, int k, List<Integer>l, int n){
        if(k == 0) {
            res.add(new ArrayList<>(l));
            return;
        }
        for(int i = start; i <= n; i++){
            l.add(i);
            addList(res, i+1, k-1, l, n);
            l.remove(l.size()-1);
        }
    }
    
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>>res = new ArrayList<>();
        for(int i = 1; i <= n-k+1; i++){
            List<Integer>l = new ArrayList<>();
            l.add(i);
            addList(res, i+1, k-1, l, n);
        }
        return res;
    }
}

File: PastYears/Practice_Algorithm/77. Combinations.cpp
class Solution {
public:
    void getV(vector<int>& v, int start, int k, int n, vector<vector<int>>&res){
        if(k == 0) {
            res.push_back(v);
            return;
        }
        for(int i = start; i <= n; i++){
            v.push_back(i);
            getV(v, i+1, k-1, n, res);
            v.pop_back();
        }
    }
    
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>>res;
        for(int i = 1; i <= n-k+1; i++){
            vector<int>tmp;
            tmp.push_back(i);
            getV(tmp, i+1, k-1, n, res);
        }
        return res;
    }
};
